docs/fabric/MAS_router/mas_router.md

```diff
@@ -81,4 +81,46 @@ Data Flow Description:
 3. The selected request is then forwarded to the appropriate output direction.
 4. Ready signals are asserted to indicate the availability for data exchange in the selected direction.
 
+- One of the main issues that we had was the HOL(head of line) problem, first lets explain the issue.
+
+### Head-of-Line Blocking (HOL):
+
+- HOL blocking happens when a packet at the head of a queue in a router or switch prevents other packets behind it from being forwarded, even if those subsequent packets are destined for different output ports.
+- Essentially, the presence of a single stalled or delayed packet can cause congestion and block the transmission of all other packets in the same queue, leading to inefficient use of network resources.
+- This situation arises because packet switching devices typically process packets in a first-in-first-out (FIFO) manner, where packets are forwarded in the order they arrive. If a packet at the front of the queue encounters a delay (e.g., due to processing or waiting for an available output port), all subsequent packets in the same queue are also delayed, regardless of their priority or urgency.
+
+### Impact of HOL:
+
+- HOL blocking can significantly degrade network performance by increasing packet latency, reducing throughput, and causing jitter.
+- In worst-case scenarios, HOL blocking can lead to packet loss and network congestion, especially in high-traffic environments.
+- It becomes more pronounced in networks with bursty traffic patterns or when there are disparities in the arrival rates of packets destined for different output ports.
+
+In our design, HOL blocking can occur if we naively insert only one fifo_arb for each router (to save space), with the output connected to other tiles. In this setup, if there is a transaction destined for the north, and behind it, there is a transaction bound for the south, the south transaction gets blocked by the north transaction. In other words, if the north tile is not ready, all other transactions will be stuck behind it.
+
+The router module that addresses the HOL blocking problem by efficiently managing the allocation of communication resources among multiple incoming requests. Here's how the router module achieves this:
+
+1. **Input Buffering**:
+   - Incoming communication requests from neighboring tiles are first buffered in FIFO buffers within the `fifo_arb` module. Each FIFO buffer corresponds to a particular communication channel or direction.
+
+2. **Arbitration**:
```

Comment:
I think it's important to note that we have a more advanced Round Robin.
We are masking the RR candidates with 2 conditions:
- There is a valid request waiting to pop (this is the common practice)
- The target of this specific request is "ready" to accept new requests.

This information should be detailed in length in the fifo_arb.
Here it should be mentioned with a pointer for more details in the fifo_arb

Reply:
done